daemon/pipeline.c

```diff
@@ -0,0 +1,372 @@
+
+#include <glib.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <gst/gst.h>
+
+#include <common.h>
+#include <modules.h>
+#include <gdbus-util.h>
+#include <log.h>
+
+#include "dbus-interface.h"
+
+static MachinelearningServicePipeline *g_gdbus_instance = NULL;
+
+static int probe_pipeline (void *data);
+static void init_pipeline (void *data);
+static void exit_pipeline (void *data);
+
+static GHashTable *pipeline_table;
+
+typedef struct _pipeline {
+  GstElement *element;
+  gint64 id;
+  GMutex lock;
+  gchar *description;
+  // Is this enough?
+  // Is register any gst related callback valid?
+  // When there occur error or state change -> should broadcast it via d-bus to clients
+} pipeline_s;
+
+// See ml_pipeline_destroy
+static void _pipeline_free (gpointer data)
+{
+  pipeline_s *p;
+
+  if (!data) {
+    g_critical ("internal error, the data should not be NULL");
+    return;
+  }
+
+  p = (pipeline_s *) data;
+  g_warning ("delete pipeline whose ID is %" G_GINT64_FORMAT, p->id);
+
+  // TODO should stop and broadcast the info to D-BUS
+  // gst_element_get_state (p->element, ...);
+  g_mutex_lock (&p->lock);
+  if (p->element)
+    gst_object_unref (p->element);
+  g_mutex_unlock (&p->lock);
+
+  g_free (p->description);
+  g_mutex_clear (&p->lock);
+
+  g_free (p);
```

Comment:
After LINE 48, ```mutex_type *newlock = &p->lock;```
do the alll clean-ups (except for g_mutex_clear).
Then you can call g_mutex_unlock & g_mutex_clear with newlock.
You may memset-zero for *p for the safety before this unlock, too.